### PR TITLE
Support lambda in aggregation functions

### PR DESCRIPTION
**Pull request by wenleix**
The code is split into two parts to help review. The first part add lambda support in `Accumulator` class, while the lambda compilation part is done in second part.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 1]
```diff
@@ -2496,14 +2502,82 @@ private AccumulatorFactory buildAccumulatorFactory(
...
+                    //
+                    // In project and filter expression compilation, ExpressionAnalyzer.getExpressionTypesFromInput
+                    // is used to generate the types of all sub-expressions. (see visitScanFilterAndProject and visitFilter)
+                    //
+                    // This does not work here since the function call representation in final aggregation node
+                    // is currently a hack: it takes intermediate type as input, and may not be a valid
+                    // function call in Presto.
+                    //
+                    // TODO: Once the intermediate representation hack is fixed, the same mechanism
+                    // in project and filter expression should be used here.
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: I think the approach you took here is pretty reasonable. I don't consider it a hack.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/LambdaBytecodeGenerator.java [hunk 2]
```diff
@@ -67,6 +77,33 @@ private LambdaBytecodeGenerator()
     {
     }
 
+    public static Map<LambdaDefinitionExpression, CompiledLambda> generateMethodsForLambda(
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/LambdaBytecodeGenerator.java]: It looks like you can unify the duplicated `generateMethodsForLambda` into this one if you add a prefix parameter here. You don't need to do this in this PR. But I think that should happen.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AccumulatorCompiler.java [hunk 3]
```diff
@@ -128,6 +129,12 @@ public static GenericAccumulatorFactoryBinder generateAccumulatorFactoryBinder(A
                 .map(StateFieldAndDescriptor::getStateField)
                 .collect(toImmutableList());
 
+        int lambdaChannelCount = metadata.getLambdaChannelInterfaces().size();
```
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/AccumulatorCompiler.java]: Removed `channel`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java (hunk 1 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: The hack is for aggregation function's intermediate representation. i.e. how the final aggregation is represented today .

File: presto-main/src/main/java/com/facebook/presto/sql/gen/LambdaBytecodeGenerator.java (hunk 2 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/LambdaBytecodeGenerator.java]: I agree. Will do this separately.

**Comment by wenleix** [comment]
Resolves https://github.com/prestodb/presto/issues/7814

**Pull request merged**